packages/entities/entities-plugins/src/components/free-form/Datakit/flow-editor/store/store.ts

```diff
@@ -0,0 +1,545 @@
+import { computed, ref } from 'vue'
+import { createInjectionState } from '@vueuse/core'
+import type {
+  ConfigNode,
+  ConfigNodeType,
+  EdgeData,
+  EdgeInstance,
+  EdgeId,
+  EditorState,
+  FieldId,
+  FieldName,
+  NodePhase,
+  NodeInstance,
+  NodeId,
+  NodeName,
+  UINode,
+} from '../../types'
+import { createId, deepClone, findFieldById } from './helpers'
+import { isImplicitName, isImplicitType } from '../node/node'
+import { initEditorState, makeNodeInstance } from './init'
+import { useValidators } from './validation'
+import { useTaggedHistory } from './history'
+
+export const [provideEditorState, useEditorState] = createInjectionState(
+  function(configNodes: ConfigNode[], uiNodes: UINode[]) {
+    const state = ref<EditorState>(initEditorState(configNodes, uiNodes))
+    const selection = ref<NodeId>()
+    const history = useTaggedHistory(state, {
+      capacity: 200,
+      clone: deepClone,
+    })
+
+    // maps
+    const nodeMapById = computed(
+      () =>
+        new Map<NodeId, NodeInstance>(
+          state.value.nodes.map((node) => [node.id, node]),
+        ),
+    )
+    const nodeMapByName = computed(
+      () =>
+        new Map<NodeName, NodeInstance>(
+          state.value.nodes.map((node) => [node.name, node]),
+        ),
+    )
+    const edgeMapById = computed(
+      () =>
+        new Map<EdgeId, EdgeInstance>(
+          state.value.edges.map((edge) => [edge.id, edge]),
+        ),
+    )
+
+    // validators bound to current maps
+    const {
+      canonicalizeConnection,
+      validateConnection,
+      validateGraph,
+      isValidConnection,
+      isValidVueFlowConnection,
+    } = useValidators(state)
+
+    // O(1) getters
+    function getNodeById(id: NodeId) {
+      return nodeMapById.value.get(id)
+    }
+    function getNodeByName(name: NodeName) {
+      return nodeMapByName.value.get(name)
+    }
+    function getEdgeById(id: EdgeId) {
+      return edgeMapById.value.get(id)
+    }
+
+    const selectedNode = computed(() =>
+      selection.value ? getNodeById(selection.value) : undefined,
+    )
+    function selectNode(id?: NodeId) {
+      selection.value = id
+    }
+
+    /* ---------- node ops ---------- */
+
+    function addNode(
+      payload: {
+        type: ConfigNodeType
+        name?: NodeName
+        phase?: NodePhase
+        position?: { x: number, y: number }
+        fields?: { input?: FieldName[], output?: FieldName[] }
+        config?: Record<string, unknown>
+      },
+      commitNow = true,
+    ) {
+      if (isImplicitType(payload.type)) {
+        console.warn('[addNode] implicit nodes are fixed.')
+        return
+      }
+      const node = makeNodeInstance({
+        type: payload.type,
+        name: payload.name,
```

Comment:
We may need a helper function to create a unique name, something like:
```ts
name: generateUniqueName(payload.name),
```

Reply:
1. Duplicate names will break the `nodeMapByName`.
2. Users might get validation error when saving, unless they set the names manually to ensure uniqueness.

Reply:
Yes, it's nice to have a function like `generateUniqueName`, but would it be better to move the usage out of `addNode`, which respects the input payload and only checks for name duplication and fails if conflicted?

Reply:
> if this `addNode` function may also be used for re-constructing nodes from a config

`addNode` and `buildNodeInstance`(out of saved config/ui data) are currently sharing a common `makeNodeInstance` function under the hood.

Should we automatically name the newly created node as something like `${type}_${countByType}`? eg. `JQ_3`, `CALL_1`, etc.

Reply:
I see. This sounds good to me

Reply:
Added a `generateNodeName` function which will be called inside `addNode` for non-implicit node types.